複数のエンドポイントを作成する追加課題

trasque opened the pull request.

trasque committed:
複数のエンドポイントを作成する追加課題

yoshi-koyama commented on README.md:
```diff
@@ -16,10 +16,18 @@ SpringWebでLocalhost:8080に `Hello, world!` を表示する。
 1. レポジトリをクローン
-2. `./gradle` でビルド
+2. `./gradle build` でビルド
 3. `./gradle bootrun` でAPサーバーを起動
```
bootRunではないかと思います。
確認お願いします！

trasque replied:
./gradlew task で確認しましたら確かにbootRunでした。
コマンドラインではbootrunで動作してしまったので気づきませんでした。
PRを修正します。

yoshi-koyama commented on src/main/java/com/raisetech/lecture06/Lecture06Controller.java:
```diff
@@ -7,8 +7,47 @@
-  public String greetings() {
-    return "Hello, world!";
+  public String hello() {
+    return "<a href=\"/bye\">ハロー！そして</a>";
```
試しにThymeleafを導入してみてもいいかも。
ただ授業ではHTMLはあまり扱わないのでちょっと触ってみるくらいでよいです。

trasque replied:
Thymeleafを追加して、@Controllerでの出力にしたいと思います。

trasque committed:
README表記、ホワイトスペース修正

- README内 bootrun -> bootRun へ修正
- hello4 エンドポイントでの全角スペース(u3000)を除去

trasque committed:
Thymeleafを導入し、テンプレート追加

- Thymeleafを導入
- テンプレート `hello.html` を追加
- @RestController -> @Controller へ変更
- コントローラーメソッドを1つにまとめた
- リクエストのパラメータによって表示内容を変化する形にした

yoshi-koyama commented on src/main/java/com/raisetech/lecture06/Lecture06Controller.java:
```diff
@@ -1,14 +1,48 @@
   @GetMapping("/hello")
-  public String greetings() {
-    return "Hello, world!";
+  public String hello(
+      @RequestParam(defaultValue = "0") String requestParam,
+      Model model) {
```
![badge](https://img.shields.io/badge/review-imo-orange.svg)
↓でもよいかと。
```suggestion
  public String hello(@RequestParam(defaultValue = "0") String requestParam, Model model) {
```

trasque replied:
横長だったので悩みましたが、引数2個だけなのでご提案の方でいきます！

yoshi-koyama commented on src/main/java/com/raisetech/lecture06/Lecture06Controller.java:
```diff
@@ -1,14 +1,48 @@
+    messageMap.put(7, "こんどこそさようなら");
+    messageMap.put(8, "おしまい");
+
+    if (requestParam.matches("[0-8]")) {
```
こちら000や888のときはどうなるか試して見てほしいです。
elseの方に行きますかね？

trasque replied:
手元で試してみました。
`requestParam=000` , `requestParam=888` でも狙った動作にはなった……と思います。
正規表現 [0-8] は、0から8までのどれか一文字、という認識です。

yoshi-koyama replied:
いえ！大丈夫そうならOKです！
正規表現はわたしも実際に試さないとわからなくて・・・。

yoshi-koyama commented on src/main/java/com/raisetech/lecture06/Lecture06Controller.java:
```diff
@@ -1,14 +1,48 @@
+    if (requestParam.matches("[0-8]")) {
+      massageIndex = Integer.valueOf(requestParam);
+      responseSetParamNumber = massageIndex.intValue();
+      responseMessage = messageMap.get(responseSetParamNumber);
```
↓でいける気がします。
```suggestion
      responseMessage = messageMap.get(Integer.valueOf(requestParam));
```

trasque replied:
ありがとうございます！！こちらに変更します。

trasque replied:
すみません！やはりこれでは動作が狙ったものになりませんでした。
`messageMap.get(0);` を呼び出すのが意図となります。
`Integer.valueOf(requestParam);` だと想定 `[0-8]` 以外の数字やアルファベットなどをキーとして使ってしまうためです。

この後の `+= 1` で Aタグに仕込む「次の番号」を用意しているので
`requestSetParamNumber = 0` で必ず「最初のMap要素」にアクセスするようにしています。

yoshi-koyama commented on src/main/java/com/raisetech/lecture06/Lecture06Controller.java:
```diff
@@ -1,14 +1,48 @@
+      @RequestParam(defaultValue = "0") String requestParam,
+      Model model) {
+    
+    String responseMessage;
```
responseMessageの宣言する箇所と実際に代入する箇所が遠いのが気になります。

trasque replied:
宣言関係は最初にまとめなきゃいけない感覚がありました。
if 郡の直前が読みやすいかなとは思いますので、移動してみます。

yoshi-koyama commented on src/main/java/com/raisetech/lecture06/Lecture06Controller.java:
```diff
@@ -1,14 +1,48 @@
+      Model model) {
+    
+    String responseMessage;
+    Integer massageIndex;
```
![badge](https://img.shields.io/badge/review-nits-green.svg)
massage -> messageですね。

trasque replied:
あああーーー！！マッサージ！！！恥ずかしい！！！なおします！！！！

yoshi-koyama commented on src/main/java/com/raisetech/lecture06/Lecture06Controller.java:
```diff
@@ -1,14 +1,48 @@
+      responseMessage = messageMap.get(responseSetParamNumber);
+    }
+
+    responseSetParamNumber += 1;
```
この`responseSetParamNumber`ってどういう意味の変数でしょうか？

yoshi-koyama replied:
意図が読み取りきれなかった・・・ 🚀

trasque replied:
一番命名に悩んだ変数でした……　英語力がポンコツです。
次のMap番号のパラメータを Thymeleaf Aタグへ差し込むための変数ですが、パッと見てその意図は伝わりづらいですよね。
nextMapParamNumberとかにしてみます。

trasque committed:
誤字訂正、宣言文位置調整、変数名変更

- massageIndex -> messageIndex
- ただし、見直しで messageIndex が不要となった
- 処理に近い位置へ宣言文を移動
- responseSetParamNumber -> nextMapParamNumber
